File: Design_HashMap.java
// Time Complexity : O(1)
// Space Complexity : O(1) Because 1000001 is a constant and it doesnt grow
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : -


// Your code here along with comments explaining your approach 
class MyHashMap {
    int[] mapArray;
    /** Initialize your data structure here. */
    public MyHashMap() {
        mapArray=new int[1000001]; // to access 0-1000000 keys we give array size 1000001
        Arrays.fill(mapArray,-1);  // assigning default value as -1 to all keys
        
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        mapArray[key]=value;
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        return mapArray[key];
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
        mapArray[key]=-1;
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: Design Hashmap.cpp
//Collision handeling - Separate chain method

// Time Complexity: O(N/K) where N is number of possible keys K is predefined buckets.
// Space Complexity: O(K+M) where M is the unique keys and K is predefined buckets.
// Run in Leetcode: Yes
// Problems: OOPS concept

class Node { 
public: 
    int key, value; 
    Node* next; 
    
}; 
Node* newNode(int key, int value){
        Node *newNode = new Node();
        newNode->key = key;
        newNode->value = value;
        newNode->next = NULL;
        return newNode;
    }

class MyHashMap {
public:
    /** Initialize your data structure here. */
    Node* head[10000];
    
    MyHashMap() {
        for(int i=0;i<10000;i++){
            head[i] = NULL;
        }
    }
    
    /** value will always be non-negative. */
    void put(int key, int value) {
        int index = key%(sizeof(head)/sizeof(head[0]));
        if(head[index]==NULL){
            head[index] = newNode(key,value);
            return;
        }
        else{
            Node* current = head[index];
            if(current->key==key){
                current->value = value;
                return;
            }
            while(current->next!=NULL){
                if(current->next->key == key){
                    current->next->value = value;
                    return;
                }
                current = current->next;
            }
            current->next = newNode(key,value);
        }
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    int get(int key) {
        int index = key%(sizeof(head)/sizeof(head[0]));
        if(head[index]==NULL){
            return -1;
        }
        else{
            Node *current = head[index];
            
            while(current!=NULL){
                if(current->key == key){
                    return current->value;
                }
                current = current->next;
            }
        }
       
        return -1;
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    void remove(int key) {
        int index = key%(sizeof(head)/sizeof(head[0]));
        if(head[index]!=NULL){
            Node* current = head[index], *prev = head[index];
            if(current->key==key){
                head[index] = current->next;
            }
            while(current!=NULL){
                if(current->key == key){
                    prev->next = current->next;
                }
                prev = current;
                current = current->next;
            }
        }
        else{
            return;
        }
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */
